datacite-commons/src/main/java/no/unit/nva/doi/handlers/ReserveDoiHandler.java

```diff
@@ -0,0 +1,137 @@
+package no.unit.nva.doi.handlers;
+
+import static java.util.Objects.nonNull;
+import static nva.commons.core.attempt.Try.attempt;
+import com.amazonaws.services.lambda.runtime.Context;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse;
+import no.unit.nva.commons.json.JsonUtils;
+import no.unit.nva.doi.model.DoiResponse;
+import no.unit.nva.doi.model.ReserveDoiRequest;
+import no.unit.nva.identifiers.SortableIdentifier;
+import no.unit.nva.model.Publication;
+import no.unit.nva.model.PublicationStatus;
+import no.unit.nva.publication.RequestUtil;
+import no.unit.nva.publication.service.impl.ResourceService;
+import nva.commons.apigateway.ApiGatewayHandler;
+import nva.commons.apigateway.RequestInfo;
+import nva.commons.apigateway.exceptions.ApiGatewayException;
+import nva.commons.apigateway.exceptions.BadGatewayException;
+import nva.commons.apigateway.exceptions.BadMethodException;
+import nva.commons.apigateway.exceptions.UnauthorizedException;
+import nva.commons.core.Environment;
+import nva.commons.core.JacocoGenerated;
+import nva.commons.core.paths.UriWrapper;
+
+public class ReserveDoiHandler extends ApiGatewayHandler<Void, DoiResponse> {
```

Comment:
Restructure code as follows:

```
package no.unit.nva.doi.handlers;

import static nva.commons.core.attempt.Try.attempt;
import com.amazonaws.services.lambda.runtime.Context;
import com.fasterxml.jackson.core.JsonProcessingException;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse;
import no.unit.nva.commons.json.JsonUtils;
import no.unit.nva.doi.model.DoiResponse;
import no.unit.nva.doi.model.ReserveDoiRequest;
import no.unit.nva.doi.validator.DoiRequestValidator;
import no.unit.nva.model.Publication;
import no.unit.nva.publication.RequestUtil;
import no.unit.nva.publication.service.impl.ResourceService;
import nva.commons.apigateway.ApiGatewayHandler;
import nva.commons.apigateway.RequestInfo;
import nva.commons.apigateway.exceptions.ApiGatewayException;
import nva.commons.apigateway.exceptions.BadGatewayException;
import nva.commons.core.Environment;
import nva.commons.core.JacocoGenerated;
import nva.commons.core.paths.UriWrapper;

public class ReserveDoiHandler extends ApiGatewayHandler<Void, DoiResponse> {

    public static final String BAD_RESPONSE_ERROR_MESSAGE = "Bad response from DataCite";
    public static final String DOI_REGISTRAR = "doi-registrar";
    protected static final String API_HOST = new Environment().readEnv("API_HOST");
    private final ResourceService resourceService;
    private final HttpClient httpClient;
    private final DoiRequestValidator validator;

    @JacocoGenerated
    public ReserveDoiHandler() {
        this(ResourceService.defaultService(), getDefaultHttpClient(), new Environment());
    }

    public ReserveDoiHandler(ResourceService resourceService,
                             HttpClient httpClient, Environment environment) {
        super(Void.class, environment);
        this.resourceService = resourceService;
        this.httpClient = httpClient;
        this.validator = new DoiRequestValidator();
    }

    public DoiResponse convertResponseToDoi(HttpResponse<String> response) throws JsonProcessingException {
        return JsonUtils.dtoObjectMapper.readValue(response.body(), DoiResponse.class);
    }

    @Override
    protected DoiResponse processInput(Void input, RequestInfo requestInfo, Context context)
        throws ApiGatewayException {
        var publication = fetchPublication(requestInfo);
        validator.validate(publication, requestInfo);
        var doiResponse = generateDoi(requestInfo.getCurrentCustomer());
        return updatePublication(doiResponse, publication);
    }

    private Publication fetchPublication(RequestInfo requestInfo) throws ApiGatewayException {
        var identifier = RequestUtil.getIdentifier(requestInfo);
        return resourceService.getPublicationByIdentifier(identifier);
    }

    @Override
    protected Integer getSuccessStatusCode(Void input, DoiResponse output) {
        return HttpURLConnection.HTTP_CREATED;
    }

    @JacocoGenerated
    private static HttpClient getDefaultHttpClient() {
        return HttpClient.newBuilder().version(HttpClient.Version.HTTP_2).build();
    }

    private DoiResponse generateDoi(URI customerId) throws BadGatewayException {
        return attempt(() -> sendRequest(customerId))
                   .map(this::convertResponseToDoi)
                   .orElseThrow(failure -> new BadGatewayException(BAD_RESPONSE_ERROR_MESSAGE));
    }

    private DoiResponse updatePublication(DoiResponse doiResponse, Publication publication) {
        var updatedPublication = publication.copy()
                                     .withDoi(doiResponse.getDoi())
                                     .build();
        resourceService.updatePublication(updatedPublication);
        return doiResponse;
    }

    private HttpResponse<String> sendRequest(URI customerId) throws IOException, InterruptedException {
        var requestBody = new ReserveDoiRequest(customerId);
        var request = HttpRequest.newBuilder()
                          .POST(BodyPublishers.ofString(JsonUtils.dtoObjectMapper.writeValueAsString(requestBody)))
                          .uri(constructUri())
                          .build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    }

    private URI constructUri() {
        return UriWrapper.fromUri(environment.readEnv(API_HOST))
                   .addChild(DOI_REGISTRAR)
                   .getUri();
    }
}
```

Reply:
Also move the methods around to make them logically follow the expected flow

Reply:
done